### PR TITLE
New: Infrastructure Policies

### PR DESCRIPTION
**Pull request by ericrpowers**
#### Description
Infrastructure policies will initially be used for security groups from AWS, but may be expanded to other cloud providers pending customer request and/or internal demand

### CODE REVIEW

**Comment by ericrpowers** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "security-groups",
    "component": "wedge",
    "pr-id": "94",
    "commit-sha": "2ccd3d800929641baa918ab971ab9ae40191a434"
  },
  {
    "project": "security-groups",
    "component": "backend",
    "pr-id": "416",
    "commit-sha": "d52328ff5874797b3c55e954db929e795f888716"
  },
  {
    "project": "security-groups",
    "component": "clad",
    "pr-id": "1137",
    "commit-sha": "ddd45d4dc42f4c2dae4fc082d407dabdfb7b8cff"
  },
  {
    "project": "security-groups",
    "component": "aws-lib",
    "pr-id": "8",
    "commit-sha": "991e7594170144a434369d229d33a8b6bf5cbfb6"
  },
  {
    "project": "security-groups",
    "component": "gaia",
    "pr-id": "334",
    "commit-sha": "7428049eda6a2e306748a06fd4b954d25efe2a70"
  },
  {
    "project": "security-groups",
    "component": "underwater",
    "pr-id": "61",
    "commit-sha": "ee3cd5d0fb8ec88ebdabbea4acdf318818006703"
  },
  {
    "project": "security-groups",
    "component": "apotests",
    "pr-id": "2022",
    "commit-sha": "2e7ab754be54d604e8405b144ae39d4be1ec477f"
  }
]
```

**Comment by ericrpowers** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "security-groups",
    "component": "underwater",
    "pr-id": "61",
    "commit-sha": "ee3cd5d0fb8ec88ebdabbea4acdf318818006703"
  },
  {
    "project": "security-groups",
    "component": "apotests",
    "pr-id": "2022",
    "commit-sha": "2e7ab754be54d604e8405b144ae39d4be1ec477f"
  },
  {
    "project": "security-groups",
    "component": "wedge",
    "pr-id": "94",
    "commit-sha": "2ccd3d800929641baa918ab971ab9ae40191a434"
  },
  {
    "project": "security-groups",
    "component": "backend",
    "pr-id": "416",
    "commit-sha": "61cc2cdacbb775e1a8a4d9474e744e446129ecf7"
  },
  {
    "project": "security-groups",
    "component": "clad",
    "pr-id": "1137",
    "commit-sha": "ddd45d4dc42f4c2dae4fc082d407dabdfb7b8cff"
  },
  {
    "project": "security-groups",
    "component": "aws-lib",
    "pr-id": "8",
    "commit-sha": "991e7594170144a434369d229d33a8b6bf5cbfb6"
  },
  {
    "project": "security-groups",
    "component": "gaia",
    "pr-id": "334",
    "commit-sha": "7428049eda6a2e306748a06fd4b954d25efe2a70"
  }
]
```

**Comment by ericrpowers** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "security-groups",
    "component": "backend",
    "pr-id": "416",
    "commit-sha": "61cc2cdacbb775e1a8a4d9474e744e446129ecf7"
  },
  {
    "project": "security-groups",
    "component": "clad",
    "pr-id": "1137",
    "commit-sha": "ddd45d4dc42f4c2dae4fc082d407dabdfb7b8cff"
  },
  {
    "project": "security-groups",
    "component": "aws-lib",
    "pr-id": "8",
    "commit-sha": "991e7594170144a434369d229d33a8b6bf5cbfb6"
  },
  {
    "project": "security-groups",
    "component": "gaia",
    "pr-id": "334",
    "commit-sha": "2d339641296dee50a285502344bd9d973b171838"
  },
  {
    "project": "security-groups",
    "component": "underwater",
    "pr-id": "61",
    "commit-sha": "ee3cd5d0fb8ec88ebdabbea4acdf318818006703"
  },
  {
    "project": "security-groups",
    "component": "apotests",
    "pr-id": "2022",
    "commit-sha": "2e7ab754be54d604e8405b144ae39d4be1ec477f"
  },
  {
    "project": "security-groups",
    "component": "wedge",
    "pr-id": "94",
    "commit-sha": "2ccd3d800929641baa918ab971ab9ae40191a434"
  }
]
```

File: doc/documentation.md [hunk 1]
```diff
@@ -6965,6 +6982,200 @@ Date of the report.
 
 Number of flows in the report.
 
+### InfrastructurePolicy
+
+Allows defining infrastructure policies to allow or prevent processing units
+identified by their tags to talk to other processing units or external services
+(also identified by their tags).
+
```
Comment (dstiliadis) [inline doc/documentation.md]: InfrastructurePolicies capture the network access rules of the underlying infrastructure and can be used to model cloud security groups, firewalls, or other ACL based mechanisms.

File: specs/infrastructurepolicy.spec [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+  rest_name: infrastructurepolicy
+  resource_name: infrastructurepolicies
+  entity_name: InfrastructurePolicy
+  package: squall
+  group: policy/networking
+  description: |-
+    Allows defining infrastructure policies to allow or prevent processing units
+    identified by their tags to talk to other processing units or external services
+    (also identified by their tags).
+  aliases:
```
Comment (dstiliadis) [inline specs/infrastructurepolicy.spec]: May be we can document it like that:

InfrastructurePolicies capture the network access rules of the underlying infrastructure and can be used to model cloud security groups, firewalls, or other ACL based mechanisms. They are not used in the identity based network authorization of Aporeto, but they can affect traffic flows in the underlying infrastructure.

File: specs/infrastructurepolicy.spec [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+  v1:
+  - name: action
+    description: Action defines the action to apply to a flow.
+    type: enum
+    exposed: true
+    allowed_choices:
+    - Allow
+    - Reject
+    - Continue
+    default_value: Allow
```
Comment (dstiliadis) [inline specs/infrastructurepolicy.spec]: Now continue for these policies.

File: specs/infrastructurepolicy.spec [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+      applyPolicyMode determines if the policy has to be applied to the
+      outgoing traffic of a PU or the incoming traffic of a PU or in both directions.
+      Default is both directions.
+    type: enum
+    exposed: true
+    allowed_choices:
+    - OutgoingTraffic
+    - IncomingTraffic
+    - Bidirectional
+    default_value: Bidirectional
```
Comment (dstiliadis) [inline specs/infrastructurepolicy.spec]: No bidirectional is possible for these policies.

File: specs/infrastructurepolicy.spec [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    parameters:
+      entries:
+      - name: mode
+        description: Matching mode.
+        type: enum
+        allowed_choices:
+        - subject
+        - object
+        default_value: object
+
```
Comment (dstiliadis) [inline specs/infrastructurepolicy.spec]: We need implementations for these relations at the backend.

File: specs/infrastructurepolicy.spec [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+    description: Returns the list of services affected by an infrastructure policy.
+    parameters:
+      entries:
+      - name: mode
+        description: Matching mode.
+        type: enum
+        allowed_choices:
+        - subject
+        - object
+        default_value: object
```
Comment (dstiliadis) [inline specs/infrastructurepolicy.spec]: This might be tricky. Leave it here .. we will have to see if we can support it.

Comment (t00f) [reply specs/infrastructurepolicy.spec]: If you're not sure, I would not add it. It is easy to add a property. Removing it or changing it requires data migration

Comment (ericrpowers) [reply specs/infrastructurepolicy.spec]: Good point. I will remove for now.

File: specs/infrastructurepolicy.spec (hunk 2 above)
Comment (ericrpowers) [reply specs/infrastructurepolicy.spec]: Thanks, added

File: specs/infrastructurepolicy.spec (hunk 4 above)
Comment (ericrpowers) [reply specs/infrastructurepolicy.spec]: Switched default to `OutgoingTraffic` when I removed

File: specs/infrastructurepolicy.spec (hunk 5 above)
Comment (ericrpowers) [reply specs/infrastructurepolicy.spec]: I want to say I piggybacked off of network access policies for these in the backend PR but will need someone to validate that.

File: specs/infrastructurepolicy.spec (hunk 6 above)
Comment (ericrpowers) [reply specs/infrastructurepolicy.spec]: Actually I want to say I did w/e network access policies did in the backend, but I will still remove until/unless we definitely want to support this.

**Comment by ericrpowers** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "security-groups",
    "component": "wedge",
    "pr-id": "94",
    "commit-sha": "2ccd3d800929641baa918ab971ab9ae40191a434"
  },
  {
    "project": "security-groups",
    "component": "backend",
    "pr-id": "416",
    "commit-sha": "7942010ab797e6d30f9fddcc1d2ed3b491ebacad"
  },
  {
    "project": "security-groups",
    "component": "aws-lib",
    "pr-id": "8",
    "commit-sha": "991e7594170144a434369d229d33a8b6bf5cbfb6"
  },
  {
    "project": "security-groups",
    "component": "gaia",
    "pr-id": "334",
    "commit-sha": "de51485a37e0df3725f0d850db326fed8ac8b74a"
  },
  {
    "project": "security-groups",
    "component": "underwater",
    "pr-id": "61",
    "commit-sha": "77c06dc9e5be240b551f8f6b1927b5818cf78476"
  },
  {
    "project": "security-groups",
    "component": "apotests",
    "pr-id": "2022",
    "commit-sha": "98dae1edc9bada77e44da8ce0e64e0a029fe4801"
  }
]
```

**Pull request merged**